[interp] add MONO_AOT_MODE_INTERP

lewurm opened the pull request:
@vargaz @kumpera @luhenry do you guys have some input on naming? how do you like `defined(HAVE_ONLINE_VES)` as a substitute for `!defined(DISABLE_JIT) || defined(ENABLE_INTERPRETER)`?

lewurm commented:
build interpreter

lewurm commented:
build interpreter

lewurm commented:
build interpreter

luhenry commented on mono/mini/aot-compiler.c:
```diff
@@ -7163,6 +7169,9 @@ mono_aot_parse_options (const char *aot_options, MonoAotOptions *opts)
 		} else if (str_begins_with (arg, "hybrid")) {
 			opts->mode = MONO_AOT_MODE_HYBRID;			
+		} else if (str_begins_with (arg, "interp")) {
+			opts->mode = MONO_AOT_MODE_FULL;
```
We should have `opts->mode = MONO_AOT_MODE_INTERP;`, remove `opts->interp = TRUE;`, add `mono_aot_mode_is_interp ()`, and check for interp mode wherever we are checking for fullAOT mode and wherever you are checking for `acfg->aot_opts.interp`.

vargaz replied:
Having a bool flag field is fine, its easier to type.

luhenry replied:
But it's not consistent. We don't use a bool for none of the other modes, and consistency and ease of understanding is way more valuable than speed of typing for something that would change very very rarely.

vargaz replied:
We do use bool for a lot of stuff, including the aot modes.

luhenry replied:
Not in this case. We use `opts->mode = MONO_AOT_MODE_(FULL|HYBRID|NORMAL)` for all the other modes.

vargaz replied:
I did suggest adding bools for that back then :)

luhenry replied:
I wasn't the one adding `opts->mode = MONO_AOT_MODE_FULL` and `mono_aot_mode_is_full`, so I only suggest we keep the same mechanism for sake of consistency and readability.

luhenry commented:
And what does ONLINE_VES mean?

kumpera commented on mono/metadata/object-offsets.h:
```diff
@@ -267,6 +267,15 @@ DECL_OFFSET(SeqPointInfo, ss_tramp_addr)
 DECL_OFFSET(SeqPointInfo, bp_addrs)
 #endif
 
+#ifdef ENABLE_INTERPRETER
```
I think this needs changes to the offset generator.

lewurm replied:
why?

kumpera commented on mono/mini/aot-runtime.c:
```diff
@@ -3313,9 +3315,11 @@ mono_aot_find_jit_info (MonoDomain *domain, MonoImage *image, gpointer addr)
 
 	nmethods = amodule->info.nmethods;
 
+#ifndef ENABLE_INTERPRETER
```
This looks wrong.
It means we can't enable the interpreter on desktop and have working (non-full)AOT.

vargaz replied:
Yes, aot-ed code will only work in the root appdomain.

lewurm replied:
ah, good catch!  would `#ifdef DISABLE_APPDOMAINS` work here?

vargaz replied:
I think the problem is that the interpreter requires some aot-ed wrappers to work right now ?
So it would not work with full-aot+domains.

lewurm replied:
yes, domains won't work with mixed mode (interp + full-aot).

the scenario is interp + domains, and that requires some aot-ed wrappers.

vargaz replied:
The wrappers will only work in the root domain, so the loading code needs to be modified somewhere to look for them in the root domain. i.e. see the 'target_domain' variable
in mono_jit_compile_method_with_opt ().

kumpera commented on mono/mini/jit.h:
```diff
@@ -54,7 +54,8 @@ typedef enum {
 	/* Same as full, but use only llvm compiled code */
-	MONO_AOT_MODE_LLVMONLY
+	MONO_AOT_MODE_LLVMONLY,
+	MONO_AOT_MODE_INTERP
```
Missing doc as the above.

lewurm commented:
@luhenry VES stands for "Virtual Execution System" and is an abbreviation often used in the .NET world.  "online" should hint that the execution engine is available at _runtime_, while an example for an "offline VES" would be the AOT mini backend.

lewurm commented:
updated PR.

@kumpera I added some documentation.

@luhenry `interp` field removed.

@vargaz can you have a look at https://github.com/mono/mono/pull/4958/commits/96c013964acc5c1f49eba9a264be5f380a7936b8 ?

lewurm commented:
build interpreter

luhenry commented on mono/mini/aot-compiler.c:
```diff
@@ -4427,7 +4427,13 @@ method_has_type_vars (MonoMethod *method)
 gboolean mono_aot_mode_is_full (MonoAotOptions *opts)
 {
-	return opts->mode == MONO_AOT_MODE_FULL;
+	return opts->mode == MONO_AOT_MODE_FULL || opts->mode == MONO_AOT_MODE_INTERP;
```
This implies that if you do `--aot=interp` then all the methods are going to be AOTted, even generics with gshared, gsharedvt, etc. I don't think that's the semantic we want from that. From my understanding interpAOT is supposed to be trampolines + wrappers that can't be interpreted.

lewurm replied:
no, that stuff is guarded by `mono_aot_mode_is_interp ()`.  Other then that, it should behave like `mono_aot_mode_is_full`.

lewurm commented:
build interpreter
